### PR TITLE
TRUNK-6341: Intelligent Locale Matching using RFC 4647

### PR DESCRIPTION
**Pull request by faraz-shamim**
TRUNK-6341: Intelligent Locale Matching using RFC 4647

## Description of what I changed
I implemented **intelligent locale matching** to replace the simple, exact-match-only logic in `InitializationFilter`.

Previously, if a user requested `fr-BE` (Belgian French) but OpenMRS only had `fr` (French), the system would default to English. This change uses **RFC 4647** (`java.util.Locale.lookup`) to properly handle:
*   **Region Fallbacks:** `fr-BE` now correctly falls back to `fr`.
*   **Quality Weights:** Headers like `fr;q=0.9, en;q=0.8` are respected.

**Specific Changes:**
1.  **`CustomResourceLoader.java`**:
    *   Added `findBestMatchLocale(String header)` to handle the parsing logic.
    *   Added a `protected` constructor to allow injecting `availableLocales`. This ensures the logic is **testable** without needing complex file system mocks (addressing previous review feedback).
2.  **`InitializationFilter.java`**:
    *   Updated to extract the `Accept-Language` header and call the new loader method.
3.  **`CustomResourceLoaderTest.java`**:
    *   Added a new test class with **19 unit tests** covering exact matches, fallbacks, quality weights, and edge cases.

## Issue I worked on
see https://issues.openmrs.org/browse/TRUNK-6341

## Checklist: I completed these to help reviewers :)
- [x] My IDE is configured to follow the [**code style**](https://openmrs.atlassian.net/wiki/spaces/docs/pages/25477044/Java+Conventions) of this project.
- [x] I have **added tests** to cover my changes.
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: web/src/main/java/org/openmrs/web/filter/initialization/InitializationFilter.java
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/filter/initialization/InitializationFilter.java:1]: Can you undo the format-only changes to this file? It makes it impossible to review this PR.

Comment (ibacher) [reply web/src/main/java/org/openmrs/web/filter/initialization/InitializationFilter.java:1]: This is why we ask you to _please_ configure your IDE to use _our_ code formatting standards.

Comment (faraz-shamim) [reply web/src/main/java/org/openmrs/web/filter/initialization/InitializationFilter.java:1]: Extremely sorry for that. Cleaned it up. Please review. Thankyou!

File: web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java [hunk 1]
```diff
@@ -67,6 +70,19 @@ private CustomResourceLoader(HttpServletRequest httpRequest) {
...
+
+	/**
+	 * Protected constructor for testing purposes that allows injecting a specific
+	 * set of available locales.
+	 * This enables unit testing of locale matching logic without file system
+	 * dependencies.
+	 *
+	 * @param availableLocales the set of locales to be used as available locales
+	 */
+	protected CustomResourceLoader(Set<Locale> availableLocales) {
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java:82]: Default visibility is probably the right thing here.

```suggestion
	CustomResourceLoader(Set<Locale> availableLocales) {
```

File: web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java [hunk 2]
```diff
@@ -67,6 +70,19 @@ private CustomResourceLoader(HttpServletRequest httpRequest) {
...
+		this.resources = new HashMap<>();
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java:83]: Since there's no sensible way to set this, just returning the empty map is probably best.

```suggestion
		this.resources = Collections.emptyMap();
```

File: web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java [hunk 3]
```diff
@@ -130,4 +146,41 @@ public Map<Locale, ResourceBundle> getResource() {
...
+	 *                             request
+	 *                             (e.g., "fr-BE,fr;q=0.9,en;q=0.8")
+	 * @return the best matching locale from available locales, or Locale.ENGLISH if
+	 *         no match is found
+	 *         or if the header is null/empty
+	 */
+	public Locale findBestMatchLocale(String acceptLanguageHeader) {
+		// Return default locale if header is null or empty
+		if (acceptLanguageHeader == null || acceptLanguageHeader.trim().isEmpty()) {
+			return Locale.ENGLISH;
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java:165]:
```suggestion
			return LocaleUtility.getDefaultLocale();
```

File: web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java [hunk 4]
```diff
@@ -130,4 +146,41 @@ public Map<Locale, ResourceBundle> getResource() {
...
+			List<LanguageRange> languageRanges = LanguageRange.parse(acceptLanguageHeader);
+
+			// Convert available locales set to list for Locale.lookup()
+			List<Locale> availableLocalesList = new ArrayList<>(getAvailablelocales());
+
+			// Use RFC 4647 lookup to find the best match
+			Locale matchedLocale = Locale.lookup(languageRanges, availableLocalesList);
+
+			// Return matched locale or default to English if no match found
+			return matchedLocale != null ? matchedLocale : Locale.ENGLISH;
```
Comment (ibacher) [inline web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java:179]:
```suggestion
			return matchedLocale != null ? matchedLocale : LocaleUtility.getDefaultLocale();
```

File: web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java (hunk 3 above)
Comment (ibacher) [reply web/src/main/java/org/openmrs/web/filter/util/CustomResourceLoader.java:165]: This may slightly disrupt the tests, but comparing `assertEquals(Locale.ENGLISH.getLanguage(), findBestMatchLocale(null).getLanguage());` as [suggested by the docs](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#getLanguage--) should work.

File: web/src/test/java/org/openmrs/web/filter/util/CustomResourceLoaderTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    @BeforeEach
+    public void setup() {
+        // Create a set of available locales for testing
+        availableLocales = new HashSet<>();
+        availableLocales.add(Locale.ENGLISH);
+        availableLocales.add(Locale.FRENCH);
+        availableLocales.add(new Locale("es")); // Spanish
+        availableLocales.add(new Locale("pt", "BR")); // Portuguese (Brazil)
+
+        // Use the protected constructor to inject test locales
```
Comment (jwnasambu) [inline web/src/test/java/org/openmrs/web/filter/util/CustomResourceLoaderTest.java:40]: Kindly as a best practice let’s remove comments that restate what the code already conveys. Clean code is easy to read and maintain.
